### PR TITLE
File handling

### PR DESCRIPTION
**Pull request by pratikbhd**
This pull request consists of the following additions:

1. List-directed writes for I/O handling
2. Handling module constructs
3. Handling private variables and functions including some other corner cases in the program implementation

### CODE REVIEW

File: tests/data/modules/test_module_09.f [hunk 1]
```diff
@@ -30,7 +30,7 @@ MODULE MYMOD9
 
       REAL, PARAMETER :: 
      &    PI = 3.14159265,
-     &    RAD=PI/180.0
+     &    RAD=0.0174532925
```
Comment (skdebray) [inline tests/data/modules/test_module_09.f]: @pratikbhd Please restore this test to the original form.  I'm not too concerned that the translator does not currently handle expressions on the RHS of a PARAMETER assignment, causing the original version of this test to fail; but I want the fact that this feature is currently not handled to be recorded somewhere so that we remember to come back and address it.  I think an easy way to remember this is to have a test input that captures the language feature that needs to be handled.

File: tests/data/modules/PETASCE_02.for [hunk 2]
```diff
@@ -194,13 +201,21 @@ C      CALL GET('SPAM', 'KCBMAX', KCBMAX)
       !Maximum crop coefficient (Kcmax) (FAO-56 Eq. 72)
       WND = MAX(1.0,MIN(WIND2m,6.0))
       CHT = MAX(0.001,CANHT)
-      SELECT CASE(MEEVP)
-        CASE('A') !Alfalfa reference
-            KCMAX = MAX(1.0,KCB+0.05)
-        CASE('G') !Grass reference
-            KCMAX = MAX((1.2+(0.04*(WND-2.0)-0.004*(RHMIN-45.0))
+
+      IF (MEEVP .EQ. 'A') THEN
```
Comment (skdebray) [inline tests/data/modules/PETASCE_02.for]: @pratikbhd Can you please create a test program that has a SELECT statement so that we know to come back and deal with it?

File: delphi/translators/for2py/scripts/translate.py [hunk 3]
```diff
@@ -258,11 +281,24 @@ def process_literal(self, root, state) -> List[Dict]:
             }
         ]
 
+    def process_io_control_spec(self, root, state) -> List[Dict]:
```
Comment (skdebray) [inline delphi/translators/for2py/scripts/translate.py]: @pratikbhd Please add a comment saying what this function does.

File: delphi/translators/for2py/scripts/translate.py [hunk 4]
```diff
@@ -392,7 +428,29 @@ def process_close(self, root, state) -> List[Dict]:
         for node in root:
             close_spec["args"] += self.parseTree(node, state)
         return [close_spec]
-        
+
+    def process_use(self, root, state) -> List[Dict]:
```
Comment (skdebray) [inline delphi/translators/for2py/scripts/translate.py]: needs comment

File: delphi/translators/for2py/scripts/translate.py [hunk 5]
```diff
@@ -392,7 +428,29 @@ def process_close(self, root, state) -> List[Dict]:
...
+        for node in root:
+            if node.tag == "only":
+                tag_spec["include"] = []
+                for item in node:
+                    if item.tag == "name":
+                        tag_spec["include"] += [item.attrib["id"]]
+
+        return [tag_spec]
+
+    def process_private_variable(self, root, state) -> List[Dict]:
```
Comment (skdebray) [inline delphi/translators/for2py/scripts/translate.py]: needs comment

File: delphi/translators/for2py/scripts/mod_index_generator.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+            return self.process_subroutine_or_program(root, state)
+
+        else:
+            status = True
+            for node in root:
+                status = self.parseTree(node, state)
+            return status
+
+    def process_subroutine_or_program(self, root, state):
+        # Initialize the list of used modules to be an empty list for each module
```
Comment (skdebray) [inline delphi/translators/for2py/scripts/mod_index_generator.py]: I think that the fact that the variable `self.useList` is being initialized to an empty list is obvious from the code; this does not need to be repeated in the comment.  What the comment *should* say is: what is the variable `self.useList` ?

File: delphi/translators/for2py/scripts/mod_index_generator.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+        for node in root:
+            status = self.parseTree(node, state)
+
+    def process_module(self, root, state):
+        module = dict()
+        module["name"] = root.attrib["name"]
+        module["source_file"] = self.fileName
+
+        # Initialize the list of used modules to be an empty list for each module
```
Comment (skdebray) [inline delphi/translators/for2py/scripts/mod_index_generator.py]: I think that the fact that the variable `self.useList` is being initialized to an empty list is obvious from the code; this does not need to be repeated in the comment. What the comment *should* say is: what is the variable `self.useList` ?

File: delphi/translators/for2py/scripts/mod_index_generator.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+        # Append the module metadata to our final ast
+        self.asts.append(module)
+
+        return status
+
+    def process_use(self, root, state):
+        self.useList.append(root.attrib["name"])
+        return True
+
+    def process_file(self, root, state):
```
Comment (skdebray) [inline delphi/translators/for2py/scripts/mod_index_generator.py]: needs comment

File: delphi/translators/for2py/scripts/mod_index_generator.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+        # At this point, the module tag has been recursively iterated to find all use statements
+        module["uses"] = self.useList
+
+        print(module)
+        # Append the module metadata to our final ast
+        self.asts.append(module)
+
+        return status
+
+    def process_use(self, root, state):
```
Comment (skdebray) [inline delphi/translators/for2py/scripts/mod_index_generator.py]: needs comment

File: delphi/translators/for2py/scripts/mod_index_generator.py [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+        if match:
+            self.fileName = match.group(1) + match.group(2)
+        for node in root:
+            self.parseTree(node, state)
+
+        return True
+
+
+
+    def analyze(self, tree: ET.ElementTree) -> List:
```
Comment (skdebray) [inline delphi/translators/for2py/scripts/mod_index_generator.py]: needs comment

File: delphi/translators/for2py/scripts/mod_index_generator.py [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+
+        status = self.parseTree(tree, ParseState())
+
+        return self.asts
+
+def get_tree(file: str) -> ET.ElementTree:
```
Comment (skdebray) [inline delphi/translators/for2py/scripts/mod_index_generator.py]: needs comment

File: delphi/translators/for2py/scripts/mod_index_generator.py [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+    return ET.parse(file).getroot()
+
+def get_index(xml_file: str):
```
Comment (skdebray) [inline delphi/translators/for2py/scripts/mod_index_generator.py]: needs comment

**Comment by skdebray** [comment]
There seem to be some build failures in Travis.  Do you know what the problem might be?

File: tests/data/modules/test_module_09.f (hunk 1 above)
Comment (pratikbhd) [reply tests/data/modules/test_module_09.f]: Test program for this construct has been in `tests/data/cases` directory

File: tests/data/modules/PETASCE_02.for (hunk 2 above)
Comment (pratikbhd) [reply tests/data/modules/PETASCE_02.for]: Test program for this construct has been in `tests/data/cases` directory

File: tests/data/modules/test_module_02.f [hunk 13]
```diff
@@ -22,9 +22,9 @@ END MODULE mymod2
       PROGRAM PGM
       USE mymod2      ! << case-insensitive
 
- 10   FORMAT (I7, I5, F8.4, F9.3)
+ 10   FORMAT (I7, I5, F8.4)
 
-      write (*,10) X, Y, PI, X*Y*PI
+      write (*,10) X, Y, PI
```
[the remark this answers is not in the document]
Comment (pratikbhd) [reply tests/data/modules/test_module_02.f]: Test program for this construct has been in `tests/data/cases` directory

File: tests/data/modules/test_module_09.f (hunk 1 above)
Comment (skdebray) [reply tests/data/modules/test_module_09.f]: Thanks.  Resolved.

File: tests/data/modules/PETASCE_02.for (hunk 2 above)
Comment (skdebray) [reply tests/data/modules/PETASCE_02.for]: Thanks.  Resolved.

**Comment by pratikbhd** [comment]
Comments added for all the above mentioned functions
